presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -284,36 +283,88 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
         checkArgument((totalBitLength <= maximumBits), format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
     }
 
+    /**
+     * The BoundedZValueRange class contains non-null Integer minimum and maximum value lists. The bounds are all inclusive.
+     */
+    private static class BoundedZValueRange
+    {
+        private final List<Integer> minimumValues;
+        private final List<Integer> maximumValues;
+
+        public BoundedZValueRange(List<Integer> minimumValues, List<Integer> maximumValues)
+        {
+            this.minimumValues = minimumValues;
+            this.maximumValues = maximumValues;
+        }
+
+        public Integer getMinimumValueAt(int index)
+        {
+            return minimumValues.get(index);
+        }
+
+        public Integer getMaximumValueAt(int index)
+        {
+            return maximumValues.get(index);
+        }
+    }
+
     private List<ZAddressRange<Long>> zOrderSearchCurve(List<ZValueRange> valueRanges)
     {
-        List<ZAddressRange<Long>> addressRanges = new ArrayList<>();
+        List<BoundedZValueRange> boundedValueRanges = fillUnspecifiedRangesWithDefaults(valueRanges);
 
-        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, fillUnspecifiedRangesWithDefaults(valueRanges), addressRanges);
+        int[] lengths = new int[valueRanges.size()];
+        for (int index = 0; index < valueRanges.size(); index++) {
+            lengths[index] = valueRanges.get(index).length;
+        }
+
+        int[] counters = new int[valueRanges.size()];
```

Comment:
If I know why "counters" is created, it will be really helpful for me to understand the code.

Reply:
After finish reading the code, I know "counters" is used to record the interval index for all the columns. Can we use intervalIndices to name it?

Reply:
`counters` is an array that helps with permuting through every combination of each column range interval. The `lengths` variable stores the maximum values in each row, while `counters` stores the current permutation.
Using the same example in the previous comment, if `lengths` is [1, 3, 2], then the valid `counters` values would be 
{[0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1], [0, 2, 0], [0, 2, 1]}.